File: android/app/src/main/kotlin/com/anitrack/app/ui/shell/LaunchIdent.kt
package com.anitrack.app.ui.shell

import android.util.Log
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshotFlow
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.TransformOrigin
import androidx.compose.ui.graphics.drawscope.clipPath
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.layout
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.imageResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.unit.Constraints
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import com.anitrack.app.BuildConfig
import com.anitrack.app.R
import com.anitrack.app.design.LocalReduceMotion
import com.anitrack.app.design.ThemeColor
import com.anitrack.app.design.brand.LaunchHandoff
import com.anitrack.app.design.brand.MarkGeometry
import com.anitrack.app.design.brand.RibbonShape
import com.anitrack.app.design.brand.drawPeriodBead
import com.anitrack.app.design.brand.drawRibbon
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.roundToInt
import kotlin.math.sin
import kotlin.math.sqrt

/*
 * THE LAUNCH — the port of `ios/Sources/App/LaunchIdent.swift`. The app icon's ribbon, at
 * cinematic scale, lit — then the app comes through it.
 *
 * Android starts every launch on the platform's own splash: the adaptive icon in a disc on the
 * canvas (`Theme.Previously.Splash`). The ident's first frame IS that picture — the same launcher
 * layers, clipped to the same disc, at the frame the exit listener reports — so the hand-off from
 * the system to the app is a picture into itself, never two splashes. Then the disc's tile
 * dissolves and the ribbon is released from it: the drawing takes over from the bitmap and grows
 * from the icon's size to a third of the screen, a travelling light crossing it as it does, the
 * icon's material forming — the ramp, the rim, the pool of warm light it casts on the canvas —
 * and the coral full stop pulses once as it settles. The composition holds, still. Once auth has
 * answered, the ident pushes through: the composition grows a hair and fades as the app emerges
 * beneath it, settling from a hair small. One gesture, one object, the brand's own material; no
 * tagline, no flash, no haptic; nothing flies into a corner.
 *
 * EVERY value here is a pure function of live time — seconds of frames actually presented, read
 * off `withFrameNanos`, with any real stall between two frames cut out of the clock. There is not
 * one animation object in this file: a `delay` counts wall time while the main thread is still
 * busy, and an animation started in the same frame as the view's appearance is folded into it.
 * The clock is a `MutableFloatState` read only inside layout, draw and layer lambdas, so a frame
 * of the ident invalidates a draw, never a composition. The ribbon is ONE fixed-size layer moved
 * by transforms — its outline (the cast shadow's) never changes and nothing is re-measured per
 * frame; a per-frame re-measure held the emulator's UI thread at 300 ms a frame.
 *
 * Under Reduce Motion (animator scale 0) nothing grows or pushes: the disc gives way to the
 * composition at rest, which holds, then crossfades to the app.
 */

// ---------------------------------------------------------------------------------------------
// The beats, in seconds of live time
// ---------------------------------------------------------------------------------------------

private object Beats {
    /** The disc's tile dissolves and the bitmap gives way to the drawing. */
    const val RELEASE = 0.25f

    /** The ribbon grows from the icon's size to the ident's — a spring with one soft settle. */
    const val GROW_AT = 0.12f
    const val GROW_RESPONSE = 0.72f
    const val GROW_DAMPING = 0.86f

    /** The light crosses the ribbon head to foot as it grows. */
    const val SWEEP_AT = 0.20f
    const val SWEEP_FOR = 0.60f

    /** The pool of light beneath arrives as the ribbon settles. */
    const val CAST_AT = 0.50f
    const val CAST_FOR = 0.40f

    /** The full stop's one pulse of light, as the light passes the tails. */
    const val PULSE_AT = 0.52f
    const val PULSE_FOR = 0.45f

    /** The composition holds before it may leave. */
    // Stated in the stop's own clock (i2-2): SWEEP_AT + PULSE_AT + 0.22, so the brand's period
    // lands on every launch, fast network or slow.
    const val HOLD_UNTIL = SWEEP_AT + PULSE_AT + 0.22f

    /**
     * The push through: the composition grows and is gone before the ground has finished lifting,
     * so it never lingers as a ghost over the screen arriving beneath it.
     */
    const val EXIT_FOR = 0.28f
    const val EXIT_SCALE = 1.12f
    const val COMPOSITION_EXIT_FRACTION = 0.65f

    /** Under Reduce Motion: the hold before the crossfade, and the crossfade. */
    const val REDUCED_HOLD = 0.6f
    const val REDUCED_FADE = 0.22f

    /** Auth's own wait is capped at 3 s; the ident does not outlast it by much. */
    const val PATIENCE = 4.0f

    /** How long the ident waits for the billboard's art after auth has answered (i2). */
    const val ART_PATIENCE = 1.6f

    /** A WALL ceiling from the first frame (i4): live seconds cut stalls out, so a starved device had none. */
    const val WALL_CEILING = 2.4f
}

private object Geometry {
    /** The ribbon's width at rest: a third of the screen, where a title card sits. */
    val RIBBON_WIDTH = 120.dp

    /** The ribbon's centre as a fraction of the screen — a hair high, a hair left, the full stop balancing it. */
    const val REST_X = 0.5f
    const val REST_Y = 0.44f
    val REST_OFFSET_X = (-14).dp

    /**
     * The system splash: the adaptive icon drawn at [ICON_BOX] with its inner two-thirds visible —
     * a [ICON_DISC] disc — centred in the window. The iOS tile (1024) maps onto that disc. Measured
     * against the exit listener's frame when it fires; predicted from the window otherwise.
     */
    val ICON_BOX = 240.dp
    val ICON_DISC = 160.dp
    const val TILE = 1024f

    /** The ribbon in the icon (design/app-icon-v2/glass3.py `geometry`): x0 = 281.6, width 376, tails at 800. */
    const val ICON_RIBBON_LEFT = 281.6f
    const val ICON_RIBBON_WIDTH = 376f
}

// ---------------------------------------------------------------------------------------------
// Easing — the unit-step response of a damped spring at rest, and two ramps
// ---------------------------------------------------------------------------------------------

/** What Compose's `spring(dampingRatio, stiffness)` traces, as a function of time. */
private fun spring(t: Float, response: Float, damping: Float): Float {
    if (t <= 0f) return 0f
    val omega = 2f * PI.toFloat() / response
    if (damping < 1f) {
        val damped = omega * sqrt(1f - damping * damping)
        return 1f - exp(-damping * omega * t) * (cos(damped * t) + (damping * omega / damped) * sin(damped * t))
    }
    return 1f - exp(-omega * t) * (1f + omega * t)
}

/** A smooth 0 → 1 over [over] seconds. */
private fun ease(t: Float, over: Float): Float {
    val x = (t / over).coerceIn(0f, 1f)
    return x * x * (3f - 2f * x)
}

/** Gathering pace (quadratic in): the exit's own curve. */
private fun easeIn(t: Float, over: Float): Float {
    val x = (t / over).coerceIn(0f, 1f)
    return x * x
}

private fun lerp(a: Float, b: Float, p: Float) = a + (b - a) * p

private fun lerp(a: Rect, b: Rect, p: Float) =
    Rect(lerp(a.left, b.left, p), lerp(a.top, b.top, p), lerp(a.right, b.right, p), lerp(a.bottom, b.bottom, p))

// ---------------------------------------------------------------------------------------------
// The frame: every value on screen at live time t, in one place
// ---------------------------------------------------------------------------------------------

private enum class Stage { Holding, Leaving, Done }

private class IdentFrame(t: Float, leaveAt: Float?, reduceMotion: Boolean) {
    val stage: Stage

    /** The bitmap disc (the system splash's picture) over the drawing, 1 → 0. */
    val disc: Float

    /** The drawing's own presence, 0 → 1, as the bitmap gives way. */
    val drawn: Float

    /** The growth from the icon's ribbon to the ident's, 0 → 1 (a hair over, then back). */
    val grow: Float

    /** The light crossing the ribbon, 0 → 1 head to foot, or `null` while there is none. */
    val sweep: Float?

    /** The pool of light beneath. */
    val cast: Float

    /** The full stop's pulse, 0 → 1 → 0. */
    val pulse: Float

    /** The push through: the composition's scale and opacity, and the ground's opacity. */
    val scale: Float
    val opacity: Float
    val ground: Float

    init {
        if (reduceMotion) {
            // The disc gives way to the composition at rest; a hold; a crossfade out. Nothing travels.
            val fadeIn = ease(t, Beats.RELEASE)
            val out = leaveAt?.let { ease(t - it, Beats.REDUCED_FADE) } ?: 0f
            disc = 1f - fadeIn
            drawn = fadeIn
            grow = 1f
            sweep = null
            cast = 1f
            pulse = 0f
            scale = 1f
            opacity = 1f - out
            ground = 1f - out
            stage = leaveAt?.let { if (t >= it + 0.3f) Stage.Done else Stage.Leaving } ?: Stage.Holding
        } else {
            val release = ease(t, Beats.RELEASE)
            disc = 1f - release
            drawn = release
            grow = spring(t - Beats.GROW_AT, Beats.GROW_RESPONSE, Beats.GROW_DAMPING)
            val pass = ease(t - Beats.SWEEP_AT, Beats.SWEEP_FOR)
            sweep = if (pass > 0f && pass < 1f) pass else null
            cast = ease(t - Beats.CAST_AT, Beats.CAST_FOR)
            val beat = ((t - Beats.PULSE_AT) / Beats.PULSE_FOR).coerceIn(0f, 1f)
            pulse = if (beat > 0f && beat < 1f) sin(beat * PI.toFloat()) else 0f
            if (leaveAt != null) {
                val since = t - leaveAt
                val out = easeIn(since, Beats.EXIT_FOR * Beats.COMPOSITION_EXIT_FRACTION)
                scale = 1f + (Beats.EXIT_SCALE - 1f) * out
                opacity = 1f - out
                ground = 1f - easeIn(since, Beats.EXIT_FOR)
                stage = if (since >= Beats.EXIT_FOR) Stage.Done else Stage.Leaving
            } else {
                scale = 1f
                opacity = 1f
                ground = 1f
                stage = Stage.Holding
            }
        }
    }
}

/**
 * Live time for the ident: seconds of frames actually presented, not wall time. A stall between
 * two frames longer than [STALL] — the main thread was busy, nothing was presented — is cut out.
 */
private class IdentClock {
    private var start = 0L
    private var last = 0L
    private var firstNanos = 0L
    private var frames = 0
    var leaveAt: Float? = null
        private set

    fun tick(nanos: Long, authReady: Boolean, artReady: Boolean, reduceMotion: Boolean): Float {
        if (start == 0L) {
            start = nanos
            last = nanos
            firstNanos = nanos
            return 0f
        }
        val gap = (nanos - last) / 1_000_000_000f
        if (gap > STALL) start += nanos - last - 16_666_667L
        last = nanos
        val t = (nanos - start) / 1_000_000_000f
        if (BuildConfig.DEBUG && ++frames % 30 == 0) Log.d("Launch", "ident t=%.2f gap=%.0fms".format(t, gap * 1000f))
        if (leaveAt == null) {
            val held = t >= if (reduceMotion) Beats.REDUCED_HOLD else Beats.HOLD_UNTIL
            val ready = authReady && (artReady || t >= Beats.ART_PATIENCE)
            val overdue = (nanos - firstNanos) / 1_000_000_000f >= Beats.WALL_CEILING
            if (held && (ready || t >= Beats.PATIENCE)) leaveAt = t else if (overdue) leaveAt = t
        }
        return t
    }

    private companion object {
        /**
         * The largest gap between two frames that still counts as continuous. A merely slow frame
         * (an emulator at 8 fps) must NOT be cut: cutting it plays the motion in stepped slow
         * motion, which reads as dropped frames on top of the drop itself.
         */
        const val STALL = 0.25f
    }
}

// ---------------------------------------------------------------------------------------------
// The screen
// ---------------------------------------------------------------------------------------------

/**
 * The ident, and its half of the hand-off.
 *
 * @param handoff the launch's shared state: auth's answer and the system icon's frame; the ident
 *   writes its phase back into it.
 * @param onLeaving the ident has started to push through: the app should begin emerging.
 * @param onFinished the ident is gone; the caller removes it.
 */
@Composable
fun LaunchIdent(
    handoff: LaunchHandoff,
    onLeaving: () -> Unit,
    onFinished: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val reduceMotion = LocalReduceMotion.current
    val clock = remember { IdentClock() }
    val time = remember { mutableFloatStateOf(0f) }
    var done by remember { mutableStateOf(false) }

    val background = ImageBitmap.imageResource(R.mipmap.ic_launcher_background)
    val foreground = ImageBitmap.imageResource(R.mipmap.ic_launcher_foreground)

    // The clock. Written inside the frame callback — which runs before this frame's layout and
    // draw — and read only by the layout, draw and layer lambdas below.
    LaunchedEffect(reduceMotion) {
        while (!done) {
            withFrameNanos { nanos ->
                time.floatValue = clock.tick(nanos, authReady = handoff.authReady, artReady = handoff.artReady, reduceMotion = reduceMotion)
            }
        }
    }

    // The hand-off, off the draw path: the phase is state the shell reads.
    LaunchedEffect(reduceMotion) {
        snapshotFlow { IdentFrame(time.floatValue, clock.leaveAt, reduceMotion).stage }.collect { stage ->
            when (stage) {
                Stage.Holding -> Unit
                Stage.Leaving -> {
                    handoff.phase = LaunchHandoff.Phase.Leaving
                    onLeaving()
                }
                Stage.Done -> {
                    done = true
                    onFinished()
                }
            }
        }
    }

    BoxWithConstraints(
        modifier = modifier
            .fillMaxSize()
            .clearAndSetSemantics {},
    ) {
        val widthPx = constraints.maxWidth.toFloat()
        val heightPx = constraints.maxHeight.toFloat()

        // --- geometry, in px, resolved once per size
        val density = LocalDensity.current
        val iconBoxPx = with(density) { Geometry.ICON_BOX.toPx() }
        val iconDiscPx = with(density) { Geometry.ICON_DISC.toPx() }
        val restWidthPx = with(density) { Geometry.RIBBON_WIDTH.toPx() }
        val restOffsetPx = with(density) { Geometry.REST_OFFSET_X.toPx() }

        fun iconBox(): Rect {
            val reported = handoff.systemIcon
            if (reported != null && reported.width > 0f) return reported
            val cx = widthPx / 2f
            val cy = heightPx / 2f
            return Rect(cx - iconBoxPx / 2f, cy - iconBoxPx / 2f, cx + iconBoxPx / 2f, cy + iconBoxPx / 2f)
        }

        /** The ribbon as the icon draws it, in px: the tile mapped onto the disc. */
        fun iconRibbon(box: Rect): Rect {
            val s = iconDiscPx / Geometry.TILE
            val tileOrigin = Offset(box.center.x - iconDiscPx / 2f, box.center.y - iconDiscPx / 2f)
            val w = Geometry.ICON_RIBBON_WIDTH * s
            val left = tileOrigin.x + Geometry.ICON_RIBBON_LEFT * s
            val top = tileOrigin.y
            return Rect(left, top, left + w, top + w * MarkGeometry.ASPECT)
        }

        val rest = run {
            val w = restWidthPx
            val h = w * MarkGeometry.ASPECT
            val cx = widthPx * Geometry.REST_X + restOffsetPx
            val cy = heightPx * Geometry.REST_Y
            Rect(cx - w / 2f, cy - h / 2f, cx + w / 2f, cy + h / 2f)
        }
        // The composition's own frame — the lockup plus room for the light it casts — so the push
        // through fades a small layer, never a full-screen one.
        val pad = rest.width * 0.5f
        val lockup = Rect(
            rest.left - pad, rest.top - pad,
            rest.left + rest.width * MarkGeometry.LOCKUP_WIDTH + pad, rest.bottom + pad,
        )

        /** Where the ribbon is at this frame: grown from the icon's to its resting size. */
        fun ribbonRect(f: IdentFrame): Rect = lerp(iconRibbon(iconBox()), rest, f.grow)

        // 1 — the ground and the system splash's disc.
        Canvas(Modifier.fillMaxSize()) {
            val f = IdentFrame(time.floatValue, clock.leaveAt, reduceMotion)
            drawRect(ThemeColor.canvas, alpha = f.ground)
            if (f.disc > 0f) {
                val box = iconBox()
                val disc = Path().apply {
                    addOval(Rect(box.center.x - iconDiscPx / 2f, box.center.y - iconDiscPx / 2f, box.center.x + iconDiscPx / 2f, box.center.y + iconDiscPx / 2f))
                }
                val offset = IntOffset(box.left.roundToInt(), box.top.roundToInt())
                val boxSize = IntSize(box.width.roundToInt(), box.height.roundToInt())
                clipPath(disc) {
                    drawImage(background, dstOffset = offset, dstSize = boxSize, alpha = f.disc)
                    drawImage(foreground, dstOffset = offset, dstSize = boxSize, alpha = f.disc)
                }
            }
        }

        // 2 — the composition: one layer sized to the lockup, which the push through scales and
        // fades. Inside it, the ribbon is its own fixed-size layer moved by transforms (its
        // outline never changes), and the full stop with its pulse is drawn beside it.
        Box(
            Modifier
                .layout { measurable, constraints ->
                    val placeable = measurable.measure(
                        Constraints.fixed(lockup.width.roundToInt().coerceAtLeast(1), lockup.height.roundToInt().coerceAtLeast(1)),
                    )
                    layout(constraints.maxWidth, constraints.maxHeight) {
                        placeable.place(lockup.left.roundToInt(), lockup.top.roundToInt())
                    }
                }
                .graphicsLayer {
                    val f = IdentFrame(time.floatValue, clock.leaveAt, reduceMotion)
                    scaleX = f.scale
                    scaleY = f.scale
                    alpha = f.opacity
                },
        ) {
            Box(
                Modifier
                    .layout { measurable, constraints ->
                        val placeable = measurable.measure(
                            Constraints.fixed(rest.width.roundToInt().coerceAtLeast(1), rest.height.roundToInt().coerceAtLeast(1)),
                        )
                        layout(constraints.maxWidth, constraints.maxHeight) {
                            placeable.place((rest.left - lockup.left).roundToInt(), (rest.top - lockup.top).roundToInt())
                        }
                    }
                    .graphicsLayer {
                        val f = IdentFrame(time.floatValue, clock.leaveAt, reduceMotion)
                        val r = ribbonRect(f)
                        val scale = r.width / rest.width
                        transformOrigin = TransformOrigin(0.5f, 0f)
                        scaleX = scale
                        scaleY = scale
                        translationX = r.center.x - rest.center.x
                        translationY = r.top - rest.top
                        alpha = f.drawn
                        shape = RibbonShape
                        shadowElevation = if (f.cast > 0.02f) rest.width * 0.14f * f.cast else 0f
                        spotShadowColor = ThemeColor.markCastShadow
                        ambientShadowColor = ThemeColor.markCastShadow
                    },
            ) {
                Canvas(Modifier.fillMaxSize()) {
                    val f = IdentFrame(time.floatValue, clock.leaveAt, reduceMotion)
                    drawRibbon(Rect(Offset.Zero, size), material = 1f, sweep = f.sweep)
                }
            }

            Canvas(Modifier.fillMaxSize()) {
                val f = IdentFrame(time.floatValue, clock.leaveAt, reduceMotion)
                if (f.drawn <= 0f) return@Canvas
                val r = ribbonRect(f)
                val bead = MarkGeometry.periodRect(Rect(r.left - lockup.left, r.top - lockup.top, r.right - lockup.left, r.bottom - lockup.top))
                if (f.pulse > 0f) {
                    val reach = bead.width * (0.7f + 1.3f * f.pulse)
                    drawCircle(
                        brush = Brush.radialGradient(
                            0f to ThemeColor.brandPeriod.copy(alpha = 0.42f), 1f to Color.Transparent,
                            center = bead.center, radius = reach,
                        ),
                        radius = reach,
                        center = bead.center,
                        alpha = f.pulse * f.drawn,
                    )
                }
                drawPeriodBead(bead.center, bead.width, lit = true)
            }
        }
    }
}

File: ios/Sources/App/LaunchIdent.swift
import SwiftUI

// The launch. The app icon's ribbon, at cinematic scale, drawn by light — then the app comes
// through it.
//
// iOS opens the app onto the bare canvas (the launch screen is `LaunchBackground`, nothing else:
// a launch image drawn from the asset catalogue is not reliably rendered — it never draws on the
// iOS 27 simulator — so the ident owns the whole entrance). On the app's first live frame a
// travelling light draws the ribbon from its head to its tails, the icon's material forming
// behind it — the ramp, the rim, the pool of warm light it casts on the canvas; as the light
// passes the tails the coral full stop lands, with one pulse of its own light. The composition
// holds, still. Then, once auth has answered, the ident pushes through: it grows a hair and fades
// as the app emerges beneath it, settling from a hair small. One gesture, one object, the brand's
// own material; no tagline, no flash, no grain, no haptic; nothing flies into a corner.
//
// EVERY value here is a pure function of live time — the seconds of frames actually presented,
// read off `TimelineView(.animation)`, with any real stall between two frames cut out of the
// clock. There are no implicit animations in the ident: an animation started in `onAppear` is
// folded into the view's first frame and never plays, and a `Task.sleep` counts wall time while
// the main thread is still busy. The ribbon is rasterised once at its size and touched only by
// a mask, a small light band and layer transforms.
//
// Under Reduce Motion nothing is drawn or pushed: the composition fades in, holds, and
// crossfades to the app.
struct LaunchIdent: View {
    @Environment(LaunchHandoff.self) private var handoff
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    /// The ident has started to push through: the app should begin emerging beneath it.
    let onLeaving: () -> Void
    /// The ident is gone.
    let onFinished: () -> Void

    @State private var clock = IdentClock()
    @State private var done = false

    /// The ribbon's width: a third of the screen, where a title card sits.
    static let ribbonWidth: CGFloat = 120
    /// The ribbon's centre as a fraction of the screen — a hair high, a hair left, the full stop
    /// balancing it.
    static let restCentre = CGPoint(x: 0.5, y: 0.44)
    static let restOffset = CGPoint(x: -14, y: 0)

    // The beats, in seconds of live time.
    /// The system's own launch transition swallows the app's first frames; the drawing waits.
    static let preroll = 0.22
    /// The light draws the ribbon head to foot: gathering, then easing into the tails.
    static let revealFor = 0.60
    /// The full stop lands as the light passes the tails.
    // 0.52 (review i2): the stop begins as the light passes ~87 % of the draw, so it has landed
    // before the composition may leave.
    static let beadAt = 0.52
    static let strike = (response: 0.40, damping: 0.72)
    /// Its one pulse of light.
    static let bloomFor = 0.45
    /// The composition holds before it may leave.
    // Stated in the bead's own clock (review i2): `preroll + beadAt + 0.22` — the bead is
    // scheduled in LIVE time (t − preroll) while the hold is in ident time, and at a flat 0.82
    // the stop began 60 ms AFTER a fast launch had started leaving, so the brand's period landed
    // only on a slow network. 0.96 now; the pulse is half through when the exit begins.
    static let holdUntil = preroll + beadAt + 0.22
    /// The picture's own patience: the ident may leave without the hero's art after this.
    static let artPatience = 1.6
    /// The push through: the composition grows and is gone before the ground has finished
    /// lifting, so it never lingers as a ghost over the screen arriving beneath it.
    static let exitFor = 0.28
    static let exitScale = 1.12
    static let compositionExitFraction = 0.50
    /// The ground lifts a beat after the composition starts leaving, so the ribbon is gone
    /// before the picture is more than half through (review i5: a translucent ribbon over art).
    static let groundLag = 0.05
    /// Auth's own wait is capped at 3 s; the ident does not outlast it by much.
    static let patience = 4.0

    /// A WALL-clock ceiling from the first frame (review i4): live seconds cut stalls out, so on
    /// a starved device the ident had no ceiling at all and held for six seconds.
    static let wallCeiling = 2.4

    var body: some View {
        GeometryReader { geo in
            TimelineView(.animation(paused: done)) { context in
                let t = clock.tick(context.date, authReady: handoff.authReady, artReady: handoff.artReady, reduceMotion: reduceMotion)
                let frame = IdentFrame(t: t, leaveAt: clock.leaveAt, reduceMotion: reduceMotion)
                stage(frame, size: geo.size)
                    .onChange(of: frame.stage) { _, stage in advance(to: stage) }
            }
        }
        .ignoresSafeArea()
        .accessibilityHidden(true)
    }

    // MARK: - The picture at time t

    private func stage(_ f: IdentFrame, size: CGSize) -> some View {
        let ribbon = Self.restRibbonRect(in: size)
        let bead = Self.restBeadRect(ribbon: ribbon)
        // The composition's own frame — the lockup plus room for the light it casts — so the push
        // through fades a small group, never a full-screen one.
        let pad = ribbon.width * 0.5
        let lockup = CGRect(x: ribbon.minX - pad, y: ribbon.minY - pad,
                            width: ribbon.width * MarkGeometry.lockupWidth + pad * 2,
                            height: ribbon.height + pad * 2)
        let ribbonLocal = CGRect(x: pad, y: pad, width: ribbon.width, height: ribbon.height)
        let beadLocal = Self.restBeadRect(ribbon: ribbonLocal)
        return ZStack {
            // The ground fades as one flat layer.
            ThemeColor.canvas.opacity(f.ground)

            ZStack(alignment: .topLeading) {
                RibbonFill(material: 1, castShadow: f.cast, width: ribbon.width, reveal: f.reveal, light: f.light)
                    .frame(width: ribbon.width, height: ribbon.height)
                    .position(x: ribbonLocal.midX, y: ribbonLocal.midY)

                // The full stop's one pulse of light as it lands.
                Circle()
                    .fill(RadialGradient(colors: [ThemeColor.brandPeriod.opacity(0.42), .clear],
                                         center: .center, startRadius: 0, endRadius: bead.width * 1.4))
                    .frame(width: bead.width * 2.8, height: bead.width * 2.8)
                    .scaleEffect(0.5 + 0.9 * f.bloom)
                    .opacity(f.bloom)
                    .position(x: beadLocal.midX, y: beadLocal.midY)
                PeriodBead(diameter: bead.width, lit: true)
                    .scaleEffect(max(f.bead, 0.001))
                    .opacity(min(f.bead * 3, 1))
                    .position(x: beadLocal.midX, y: beadLocal.midY)
            }
            .frame(width: lockup.width, height: lockup.height)
            // The push through: the composition grows a hair and fades.
            .scaleEffect(f.scale)
            .opacity(f.opacity)
            .position(x: lockup.midX, y: lockup.midY)
        }
        .frame(width: size.width, height: size.height)
    }

    // MARK: - The composition at rest

    static func restRibbonRect(in size: CGSize) -> CGRect {
        let w = ribbonWidth
        let h = ribbonWidth * MarkGeometry.aspect
        let cx = size.width * restCentre.x + restOffset.x
        let cy = size.height * restCentre.y + restOffset.y
        return CGRect(x: cx - w / 2, y: cy - h / 2, width: w, height: h)
    }

    static func restBeadRect(ribbon: CGRect) -> CGRect {
        let d = ribbon.width * MarkGeometry.periodDiameter
        let cx = ribbon.minX + ribbon.width * MarkGeometry.periodCenterX
        let cy = ribbon.minY + ribbon.width * MarkGeometry.periodCenterY
        return CGRect(x: cx - d / 2, y: cy - d / 2, width: d, height: d)
    }

    // MARK: - The hand-off

    private func advance(to stage: IdentFrame.Stage) {
        switch stage {
        case .holding:
            break
        case .leaving:
            handoff.phase = .leaving
            onLeaving()
        case .done:
            done = true
            onFinished()
        }
    }
}

/// Live time for the ident: seconds of frames actually presented, not wall time.
///
/// A reference type on purpose — it is written from inside the timeline's body, where a state
/// write is not allowed, and nothing needs to re-render because of it: the tick that writes it
/// is the tick that reads it.
@MainActor
final class IdentClock {
    private var start: Date?
    private var last: Date?
    /// Live time at which the ident was cleared to leave.
    private(set) var leaveAt: Double?
    private var firstFrame: Date?

    /// The largest gap between two frames that still counts as continuous. Anything longer is a
    /// stall — the main thread was busy, no frame was presented — and is cut out of the clock. A
    /// merely slow frame (the simulator at 8–10 fps) must NOT be cut: cutting it plays the motion
    /// in stepped slow motion.
    private static let stall: TimeInterval = 0.25

    func tick(_ date: Date, authReady: Bool, artReady: Bool = true, reduceMotion: Bool) -> Double {
        guard let start else {
            self.start = date
            last = date
            return 0
        }
        // Stalls are cut out only while HOLDING (review i5): once leaving, wall time drives the
        // exit, so a busy main thread shortens the ghost instead of freezing it mid-fade.
        if leaveAt == nil, let last, date.timeIntervalSince(last) > Self.stall {
            self.start = start.addingTimeInterval(date.timeIntervalSince(last) - 1.0 / 60.0)
        }
        last = date
        let t = date.timeIntervalSince(self.start ?? date)
        if firstFrame == nil { firstFrame = date }
        if leaveAt == nil {
            let held = t >= (reduceMotion ? 0.6 : LaunchIdent.holdUntil)
            let overdue = date.timeIntervalSince(firstFrame ?? date) >= LaunchIdent.wallCeiling
            // The app emerges beneath the ident WITH its picture (review i2: it used to come
            // through as a tint, then a blur, then the poster — three arrivals in the first
            // second); auth's patience stays the ceiling.
            let ready = authReady && (artReady || t >= LaunchIdent.artPatience)
            if held, ready || t >= LaunchIdent.patience { leaveAt = t } else if overdue { leaveAt = t }
        }
        return t
    }
}

/// Every value on screen at live time `t`, in one place.
struct IdentFrame: Equatable {
    enum Stage: Equatable { case holding, leaving, done }

    let stage: Stage
    /// The ribbon drawn head to foot, 0 → 1.
    let reveal: Double
    /// Where the drawing light is along the ribbon while it draws, or `nil` once it has.
    let light: Double?
    /// The pool of light beneath, forming with the ribbon.
    let cast: Double
    /// The full stop's strike, 0 → 1 (one restrained overshoot).
    let bead: Double
    /// Its pulse of light, 0 → 1 → 0.
    let bloom: Double
    /// The push through: the composition's scale and opacity, and the ground's opacity.
    let scale: Double
    let opacity: Double
    let ground: Double

    init(t: Double, leaveAt: Double?, reduceMotion: Bool) {
        if reduceMotion {
            // A crossfade in, a hold, a crossfade out. Nothing draws or pushes.
            let fadeIn = IdentFrame.ease(t, over: 0.12)
            let out = leaveAt.map { IdentFrame.ease(t - $0, over: 0.22) } ?? 0
            reveal = 1
            light = nil
            cast = 1
            bead = 1
            bloom = 0
            scale = 1
            opacity = fadeIn * (1 - out)
            ground = 1 - out
            stage = leaveAt.map { t >= $0 + 0.3 ? .done : .leaving } ?? .holding
            return
        }
        let live = t - LaunchIdent.preroll
        let drawn = IdentFrame.ease(live, over: LaunchIdent.revealFor)
        reveal = drawn
        light = drawn > 0 && drawn < 1 ? drawn : nil
        // The pool of light arrives as the ribbon finishes: it lights the page it hangs on.
        cast = IdentFrame.ease(live - LaunchIdent.revealFor * 0.75, over: 0.40)
        bead = IdentFrame.spring(live - LaunchIdent.beadAt, response: LaunchIdent.strike.response,
                                 damping: LaunchIdent.strike.damping)
        let pulse = min(max((live - LaunchIdent.beadAt) / LaunchIdent.bloomFor, 0), 1)
        bloom = pulse > 0 && pulse < 1 ? sin(pulse * .pi) : 0
        if let leaveAt {
            let since = t - leaveAt
            let out = IdentFrame.easeIn(since, over: LaunchIdent.exitFor * LaunchIdent.compositionExitFraction)
            scale = 1 + (LaunchIdent.exitScale - 1) * out
            opacity = 1 - out
            ground = 1 - IdentFrame.easeIn(since - LaunchIdent.groundLag, over: LaunchIdent.exitFor)
            stage = since >= LaunchIdent.exitFor + LaunchIdent.groundLag ? .done : .leaving
        } else {
            scale = 1
            opacity = 1
            ground = 1
            stage = .holding
        }
    }

    /// The unit-step response of a damped spring at rest — what SwiftUI's
    /// `spring(response:dampingFraction:)` traces — as a function, so it can be read at any live
    /// time rather than driven by the animation system.
    static func spring(_ t: Double, response: Double, damping: Double) -> Double {
        guard t > 0 else { return 0 }
        let omega = 2 * Double.pi / response
        if damping < 1 {
            let damped = omega * (1 - damping * damping).squareRoot()
            return 1 - exp(-damping * omega * t) * (cos(damped * t) + (damping * omega / damped) * sin(damped * t))
        }
        return 1 - exp(-omega * t) * (1 + omega * t)
    }

    /// A smooth 0 → 1 over `over` seconds.
    static func ease(_ t: Double, over duration: Double) -> Double {
        let x = min(max(t / duration, 0), 1)
        return x * x * (3 - 2 * x)
    }

    /// Fast, then easing into place (cubic out).
    static func easeOut(_ t: Double, over duration: Double) -> Double {
        let x = min(max(t / duration, 0), 1)
        return 1 - pow(1 - x, 3)
    }

    /// Gathering pace (quadratic in): the exit's own curve.
    static func easeIn(_ t: Double, over duration: Double) -> Double {
        let x = min(max(t / duration, 0), 1)
        return x * x
    }
}
